src/librustc_passes/check_const.rs

```diff
@@ -135,15 +152,47 @@ impl<'tcx> CheckConstVisitor<'tcx> {
 
         let const_kind = self.const_kind
             .expect("`const_check_violated` may only be called inside a const context");
-
         let msg = format!("`{}` is not allowed in a `{}`", expr.name(), const_kind);
-        match expr {
-            | NonConstExpr::Match(hir::MatchSource::Normal)
-            | NonConstExpr::Match(hir::MatchSource::IfDesugar { .. })
-            | NonConstExpr::Match(hir::MatchSource::IfLetDesugar { .. })
-            => feature_err(&self.tcx.sess.parse_sess, sym::const_if_match, span, &msg).emit(),
 
-            _ => span_err!(self.tcx.sess, span, E0744, "{}", msg),
+        let gates = gates.unwrap_or(&[]);
+        let missing_gates: Vec<_> = gates
+            .iter()
+            .copied()
+            .filter(|&g| !features[g])
+            .collect();
+
+        match missing_gates.as_slice() {
+            &[] => span_err!(self.tcx.sess, span, E0744, "{}", msg),
+
+            // If the user enabled `#![feature(const_loop)]` but not `#![feature(const_if_match)]`,
+            // explain why their `while` loop is being rejected.
+            &[gate @ sym::const_if_match] if gates.contains(&sym::const_loop) => {
+                let mut err = feature_err(&self.tcx.sess.parse_sess, gate, span, &msg);
```

Comment:
```suggestion
                feature_err(&self.tcx.sess.parse_sess, gate, span, &msg)
```
Use chaining wherever possible when working with diagnostic builders.